topdown/json.go

```diff
@@ -5,12 +5,116 @@
 package topdown
 
 import (
+	"fmt"
+	"strconv"
 	"strings"
 
 	"github.com/open-policy-agent/opa/ast"
 	"github.com/open-policy-agent/opa/topdown/builtins"
 )
 
+func builtinJSONRemove(_ BuiltinContext, operands []*ast.Term, iter func(*ast.Term) error) error {
+
+	// Expect an object and a string or array/set of strings
+	obj, err := builtins.ObjectOperand(operands[0].Value, 1)
+	if err != nil {
+		return err
+	}
+
+	// Build a list of json pointers to remove
+	paths, err := getJSONPaths(operands[1].Value)
+	if err != nil {
+		return err
+	}
+
+	newObj, err := jsonRemove(ast.NewTerm(obj), ast.NewTerm(pathsToObject(paths)))
+	if err != nil {
+		return err
+	}
+
+	return iter(newObj)
+}
+
+// jsonRemove returns a new term that is the result of walking
+// through a and omitting removing any values that are in b but
+// have ast.Null values (ie leaf nodes for b).
+func jsonRemove(a *ast.Term, b *ast.Term) (*ast.Term, error) {
+	if a == nil {
+		// Shouldn't happen for well formed ast.Value's, just bubble the nil value back up
+		return nil, nil
+	}
+
+	if b == nil {
+		// The paths diverged, return a
+		return a.Copy(), nil
+	}
+
+	var bObj ast.Object
+	switch bValue := b.Value.(type) {
+	case ast.Object:
+		bObj = bValue
+	case ast.Null:
+		// Means we hit a leaf node on "b", dont add the value for a
+		return nil, nil
+	default:
+		// The paths diverged, return a
+		return a.Copy(), nil
+	}
+
+	switch aValue := a.Value.(type) {
+	case ast.String, ast.Number, ast.Boolean, ast.Null:
+		return a.Copy(), nil
```

Comment:
Can we omit the copy here?

Reply:
Huh, yea I guess we can for all of these. I was thinking we needed to make a full copy but I guess it doesn't hurt to reference the original values since they shouldn't change.